src/gui/desktopintegration.cpp

```diff
@@ -0,0 +1,270 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "desktopintegration.h"
+
+#include <chrono>
+
+#include <QMenu>
+#include <QTimer>
+
+#ifndef Q_OS_MACOS
+#include <QSystemTrayIcon>
+#endif
+
+#include "base/logger.h"
+#include "base/preferences.h"
+#include "uithememanager.h"
+
+#ifdef Q_OS_MACOS
+#include "macutilities.h"
+#endif
+
+#ifdef QBT_USES_CUSTOMDBUSNOTIFICATIONS
+#include "notifications/dbusnotifier.h"
+#endif
+
+namespace
+{
+#ifdef Q_OS_MACOS
+    DesktopIntegration *desktopIntegrationInstance = nullptr;
+
+    bool handleDockClicked(id self, SEL cmd, ...)
+    {
+        Q_UNUSED(self);
+        Q_UNUSED(cmd);
+
+
+        Q_ASSERT(desktopIntegrationInstance);
+        emit desktopIntegrationInstance->activationRequested();
+
+        return true;
+    }
+#endif
+}
+
+using namespace std::chrono_literals;
+
+#define SETTINGS_KEY(name) u"GUI/" name
+#define NOTIFICATIONS_SETTINGS_KEY(name) (SETTINGS_KEY(u"Notifications/"_qs) name)
+
+DesktopIntegration::DesktopIntegration(QObject *parent)
+    : QObject(parent)
+    , m_storeNotificationEnabled {NOTIFICATIONS_SETTINGS_KEY(u"Enabled"_qs)}
+#ifdef QBT_USES_CUSTOMDBUSNOTIFICATIONS
+    , m_storeNotificationTimeOut {NOTIFICATIONS_SETTINGS_KEY(u"Timeout"_qs)}
+#endif
+{
+#ifndef Q_OS_MACOS
+    if (Preferences::instance()->systemTrayEnabled())
+        createTrayIcon(20);
+#else
+    desktopIntegrationInstance = this;
+    MacUtils::overrideDockClickHandler(handleDockClicked);
+#endif
+
+#ifdef QBT_USES_CUSTOMDBUSNOTIFICATIONS
+    if (isNotificationsEnabled())
+    {
+        m_notifier = new DBusNotifier(this);
+        connect(m_notifier, &DBusNotifier::messageClicked, this, &DesktopIntegration::notificationClicked);
+    }
+#endif
+
+    connect(Preferences::instance(), &Preferences::changed, this, &DesktopIntegration::onPreferencesChanged);
+}
+
+DesktopIntegration::~DesktopIntegration()
+{
+}
+
+bool DesktopIntegration::isActive() const
+{
+#ifndef Q_OS_MACOS
+    return (m_systrayIcon != nullptr);
+#endif
+    return true;
```

Comment:
There should be an `#else` otherwise it results in double return which doesn't make sense.
Also in `DesktopIntegration::notificationTimeout()`.

Also I would suggest using `#ifdef Q_OS_MACOS ... #else ...` globally so it clearly shows macos has a different code path than others.